### PR TITLE
add gemmx prameter support

### PR DESCRIPTION
**Pull request by xiaoling-yi**
In this PR, we add the parameter support for gemmx. More specifically, we:
* set the gemmx pe array size, meshRow, tileSize, and meshCol as configurable parameters in the cfg file. Now we can set these parameters to different values, but please take care to modify the streamer parameters too! Just so you know, the transpose extension only works for 512 data width for readers.
* Note: the pipelined SIMD only works for meshRow * meshCol = 64 cases so if you set pe array size that doesn't have 64 output, please take care to set with_pipeline = false.
* modify software test and datagen.py to accommodate arbitrary pe array size. Now only fixed the matmul test. Conv test modification will be done later.
* relevant modification in snax-gen.py. The format tool did some format, so it looks like many changes but only very minimal changes actually.

### CODE REVIEW

**Review by rgantonio — APPROVED** [review]
OK! Would it be more useful though that there is a default setting for the GeMM config?

Otherwise, people who want to use your gemm needs to always set the size?

File: hw/chisel_acc/src/main/scala/snax_acc/gemm_simd/BlockGemmRescaleSIMD.scala [hunk 1]
```diff
@@ -122,3 +122,62 @@ object BlockGemmRescaleSIMD extends App {
...
+          case Array(key, value) if key.startsWith("--") => key.drop(2) -> value
+        }
+        .toMap
+      if (parsed_args.size != 4) {
+        throw new Exception(
+          "Please provide the meshRow, meshCol, tileSize, and withPipeline. Example usage: sbt 'runMain snax_acc.gemmx.BlockGemmRescaleSIMDGen --meshRow 2 --meshCol 2 --tileSize 16 --withPipeline true'"
+        )
+      }
+      parsed_args
+    }
```
[the remark this answers is not in the document]
Comment (xiaoling-yi) [reply hw/chisel_acc/src/main/scala/snax_acc/gemm_simd/BlockGemmRescaleSIMD.scala:142]: Yes! I will do that.

**Comment by xiaoling-yi** [comment]
> Would it be more useful though that there is a default setting for the GeMM config?

I don't know. Maybe it's better to ensure that people explicitly set the PE size?

**Comment by rgantonio** [comment]
> I don't know. Maybe it's better to ensure that people explicitly set the PE size?
> The streamer parameters should be matched with the PE array size

I'm not too picky so okay haha. Maybe we can add somekind of documentation on our documentation page for this 😄

**Comment by xiaoling-yi** [comment]
> > I don't know. Maybe it's better to ensure that people explicitly set the PE size?
> > The streamer parameters should be matched with the PE array size
> 
> I'm not too picky so okay haha. Maybe we can add somekind of documentation on our documentation page for this 😄

We can!

**Review by JosseVanDelm — COMMENTED** [review]
Hi, nice PR!

already a few high-level questions.
I'm not a big fan of our current build process because it routes everything to a single python command that can not be pipelined, and which parallel make does not like https://github.com/KULeuven-MICAS/snax_cluster/issues/228

For this specific PR though:
> Just so you know, the transpose extension only works for 512 data width for readers.

> Note: the pipelined SIMD only works for meshRow * meshCol = 64 cases so if you set pe array size that doesn't have 64 output, please take care to set with_pipeline = false.

Can you please please please verify this somewhere in software instead of just this comment

Who sets `TRANSPOSE_EXTENSION_ENABLE` in the software makefile?
Is there an automatic way this is set?

**Comment by xiaoling-yi** [comment]
> For this specific PR though:
> 
> > Just so you know, the transpose extension only works for 512 data width for readers.
> 
> > Note: the pipelined SIMD only works for meshRow * meshCol = 64 cases so if you set pe array size that doesn't have 64 output, please take care to set with_pipeline = false.
> 
> Can you please please please verify this somewhere in software instead of just this comment
> 

I verified locally. Later relevant documentation will be added to push a warning on this.

> Who sets `TRANSPOSE_EXTENSION_ENABLE` in the software makefile? Is there an automatic way this is set?

The streamer header file gen will set it if a transpose extension is instantiated.

**Pull request merged**